### PR TITLE
Implement the initial version of pubsub feature

### PR DESCRIPTION
**Pull request by NisalaNiroshana**
## Purpose
Introduce initial implementation of pubsub feature for dashboard component.

## Goals
Users can introduce communication between widgets in the dashboard with this pubsub feature.

### CODE REVIEW

File: components/dashboards-web-component/src/designer/DashboardDesigner.jsx [hunk 1]
```diff
@@ -185,6 +193,33 @@ export default class DashboardDesigner extends Component {
         );
     }
 
+
+    handleWidgetConfiguration(event, isSameWidgetClicked) {
+        if (isSameWidgetClicked) {
+            this.setState({widgetConfigPanelOpen: false}, this.handleDashboardContainerStyles);
```
Comment (lasanthaS) [inline components/dashboards-web-component/src/designer/DashboardDesigner.jsx]: Can't refactor this as following ?
```
 this.setState({widgetConfigPanelOpen: !isSameWidgetClicked}, this.handleDashboardContainerStyles);
```

File: components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                this.unWirePubSubInDashboardJSON(dashboardLayout.toConfig().content,
+                    selectedWidget.config.content[0].props.id, event.currentTarget.id);
+                this.setState({checked: false});
+            }
+        }
+    }
+
+    wirePubSubInDashboardJSON(content, subscriberId, publisherId) {
+        content.forEach(contentItem => {
+            if (!(contentItem.type === 'component')) {
```
Comment (lasanthaS) [inline components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx]: Can simplify this as 

```
(contentItem.type !== 'component')
```

File: components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    constructor(props) {
+        super(props);
+        this.getPublishers = this.getPublishers.bind(this);
+        this.onCheck = this.onCheck.bind(this);
+        this.state = {
+            checked: false
+        }
+    }
+
+    onCheck(event, isInputChecked) {
```
Comment (TanyaM) [inline components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx]: Let's rename this to a meaningful name

File: components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                                          onCheck={this.onCheck}
+                                          checked={this.state.checked}
+                                          className="publishers-list"/>)
+            }
+        }
+        return publishers;
+    }
+
+    render() {
+        return <Drawer open={this.props.open} openSecondary={true}
```
Comment (lasanthaS) [inline components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx]: Better to wrap this with parentheses (refer: https://github.com/airbnb/javascript/blob/master/react/README.md#parentheses)

File: components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        } else {
+            if (pubsubComponent.unwire(selectedWidget.config.content[0].props.id, event.currentTarget.id)) {
+                this.unWirePubSubInDashboardJSON(dashboardLayout.toConfig().content,
+                    selectedWidget.config.content[0].props.id, event.currentTarget.id);
+                this.setState({checked: false});
+            }
+        }
+    }
+
+    wirePubSubInDashboardJSON(content, subscriberId, publisherId) {
```
Comment (TanyaM) [inline components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx]: let's rename as persistePubSubWiringInDashboardJSON

File: components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                    if (!contentItem.props.configs.pubsub.publishers) {
+                        contentItem.props.configs.pubsub.publishers = [];
+                    }
+                    contentItem.props.configs.pubsub.publishers.push(publisherId.toString())
+                }
+            }
+        });
+    }
+
+    unWirePubSubInDashboardJSON(content, subscriberId, publisherId) {
```
Comment (TanyaM) [inline components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx]: same as above

File: components/dashboards-web-component/src/designer/components/WidgetsList.jsx [hunk 7]
```diff
@@ -81,6 +83,17 @@ export default  class WidgetsList extends React.Component {
 
     initializeWidgetList(isWidgetsLoaded, initDashboardFlag) {
         let newItemConfig;
+        var generateguid = function () {
```
Comment (lasanthaS) [inline components/dashboards-web-component/src/designer/components/WidgetsList.jsx]: Better move this to a util class.

File: components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx (hunk 2 above)
Comment (lasanthaS) [reply components/dashboards-web-component/src/designer/components/WidgetConfigurationPanel.jsx]: Similar conditions found down the line. Apply the same for them as well.

File: components/dashboards-web-component/src/designer/components/WidgetsList.jsx (hunk 7 above)
Comment (TanyaM) [inline components/dashboards-web-component/src/designer/components/WidgetsList.jsx]: Let's move this to a util file

File: components/dashboards-web-component/src/utils/PubSubComponent.jsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+    unwire(subscriberId, publisherId) {
+        if (this.pubsubCallbackMap.get(subscriberId + "_" + publisherId)) {
+            dashboardLayout.eventHub.off(publisherId, this.pubsubCallbackMap.get(subscriberId + "_" + publisherId));
+            return true;
+        }
+        return false;
+    }
+
+    addPublisher(generatedPublisherName, id) {
```
Comment (TanyaM) [inline components/dashboards-web-component/src/utils/PubSubComponent.jsx]: rename to denote that function is adding publisher to map

File: components/dashboards-web-component/src/utils/PubSubComponent.jsx [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        this.publishersList.set(generatedPublisherName, id);
+    }
+
+    getPublishersMap() {
+        return this.publishersList;
+    }
+
+    addWidgetToPublishersList(widget) {
```
Comment (TanyaM) [inline components/dashboards-web-component/src/utils/PubSubComponent.jsx]: Let's not have 2 functions saying addPublisher and addWidgetToPublisherList.
Rename this to isPublisherWidget

**Pull request merged**